### PR TITLE
Work relations

### LINKED ISSUE #110: Work relations
- WR on salary edit page, remove other unnecessary fields.
- Work relation should replace the employee link in altinnReport1 and altinnReport1Employee tables.
- For one employee have a check that there are no WR for the same virksomhet at the same time.
- New DB table for WR and migrate old data.
- Remove old fields that are no longer needed in DB.
- Change report generating so it picks up WR not employee only.

This is what fields are needed for WR on employee.edit page(a row for each WR):

| Ansatt ved | Arbeid start | Arbeid slutt | Yrke | Samme posisjon siden | ... |
| --- | --- | --- | --- | --- | --- |
| < Virksomhet > | 0000-00-00 | 0000-00-00 | < Yrke > | 0000-00-00 | ... |

| ... | Arbeidstid | Skifttype | Ansettelsestype | Stillingsprosent | ... |
| --- | --- | --- | --- | --- | --- |
| ... | < Arbeidstid > | < Skifttype > | < Ansettelsestype > | % | ... |

| ... | Stillingsprosent oppdatert | Timer hver uke ved full stilling | Kommune |
| --- | --- | --- | --- |
| ... | 0000-00-00 | # | < Kommune > |

### PR DESCRIPTION
**Pull request by mladjo2505**
Closes #110

### CODE REVIEW

File: db/changes/128_add_work_relation.sql [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+TypeOfEmployment varchar(50) DEFAULT 'ordinaertArbeidsforhold',
+WorkPercent decimal(16,2) DEFAULT 0,
+WorkPercentUpdatedAt date,
+WorkMeasurement decimal(4,2) DEFAULT 0,
+TS timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+PRIMARY KEY (WorkRelationID)
+);
+
+-- Migrate existing information to new table for work relations
+INSERT INTO workrelation(AccountPlanID, OccupationID, SubcompanyID, KommuneID, WorkStart, WorkStop, InCurrentPositionSince, WorkTimeScheme, ShiftType, TypeOfEmployment, WorkPercent, WorkMeasurement)
```
Comment (mladjo2505) [inline db/changes/128_add_work_relation.sql]: Add WorkPercentUpdatedAt

File: db/changes/128_add_work_relation.sql [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+INSERT INTO altinnReport1WorkRelation(AltinnReport1ID, WorkRelationID)
+SELECT ar1e.AltinnReport1ID, wr.WorkRelationID
+FROM altinnReport1employee ar1e JOIN accountplan a ON ar1e.AccountPlanID = a.AccountPlanID JOIN workrelation wr ON wr.AccountPlanID = a.AccountPlanID ORDER BY wr.WorkRelationID LIMIT 1;
+
+-- Add work relartion foreign key to salary table
+ALTER TABLE salary
+ADD WorkRelationID int(11);
+
+-- Populate the newly created WorkRelationID field based on eployee for the salary
+UPDATE salary
```
Comment (mladjo2505) [inline db/changes/128_add_work_relation.sql]: Not linking to SalaryID, fix this

File: modules/altinnsalary/model/altinnreport.class.php [hunk 3]
```diff
@@ -337,14 +339,15 @@ function setLoennOgGodtgjoerelse($salary, $code_for_tax_calculation, &$loennOgGo
...
-  function generateArbeidsforholdArray($salary, $employee, $use_only_employee_info = false) {
+  function generateArbeidsforholdArray($salary, $employee, $work_relation, $use_only_employee_info = false) {
     global $_lib;
 
     // get the occupation of the employee in the company
     if ($use_only_employee_info) {
       // Error is: Occupation not set for salary employee ' . self::fullNameForErrorMessage($employee)
-      self::checkIfEmpty($employee->OccupationID, 'Ansatt: Mangler yrke p&aring; ' . self::fullNameForErrorMessage($employee));
-      $occupation_id = $employee->OccupationID;
+      // TODO: Update error messages
```
Comment (mladjo2505) [inline modules/altinnsalary/model/altinnreport.class.php]: Do this

**Comment by MrdjanStojimirovic** [comment]
First round review: @MrdjanStojimirovic
Final review with: @stoivo

**Comment by stoivo** [comment]
- [x] Line up salary second header line.
- [x] t=altinnsalary.altinn1list select period with no salaries, displays sql warning.
- [x] When the WR is selected on salary, if we change something on the WR the update button does not update the salary, until you select a different WR and back.
- [x] Oppdater arbeidsforhold => Endre arbeidsforhold
    (Update  WR   => change WR)
- [x] In t=altinnsalary.altinn1list have two text area fields, one for xml and one for debug info.

File: code/lib/form/form_elements3.class.php [hunk 4]
```diff
@@ -1131,7 +1131,13 @@ function accountplan_number_menu($args) {
             $disabled = "";
         }
 
-        $element = "<select $disabled $selectedcolor name=\"".$name."\" tabindex=\"".$args['tabindex']."\" accesskey=\"".$args['accesskey']."\" class=\"".$args['class']."\"" . $element;
+        if(isset($args['onchange'])) {
+            $onchange = $args['onchange'];
```
Comment (stoivo) [inline code/lib/form/form_elements3.class.php]: Change to

``` php
$onchange = " onchange=\"" . $args['onchange'] . "\"";
```

File: db/changes/128_add_work_relation.sql [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+-- the employee it belongs to
+INSERT INTO altinnReport1WorkRelation(AltinnReport1ID, WorkRelationID)
+SELECT ar1e.AltinnReport1ID, wr.WorkRelationID
+FROM altinnReport1employee ar1e JOIN accountplan a ON ar1e.AccountPlanID = a.AccountPlanID JOIN workrelation wr ON wr.AccountPlanID = a.AccountPlanID ORDER BY wr.WorkRelationID LIMIT 1;
+
+-- Add work relartion foreign key to salary table
+ALTER TABLE salary
+ADD WorkRelationID int(11);
+
+-- Populate the newly created WorkRelationID field based on eployee for the salary
```
Comment (stoivo) [inline db/changes/128_add_work_relation.sql]: eployee =>  employee

File: modules/altinnsalary/model/altinnreport.class.php [hunk 6]
```diff
@@ -582,6 +584,16 @@ function queryStringForCurrentlyEmployedEmployees() {
   }
 
 /* Helper function that generates the query to get
+ * the included work relations
+ */
+  function queryStringForIncludedWorkRelations() {
+    $query_work_relations = "SELECT wr.*
+                             FROM workrelation wr
+                             WHERE wr.WorkRelationID IN (" . implode($this->work_relation_ids, ', ') . ")";
```
Comment (stoivo) [inline modules/altinnsalary/model/altinnreport.class.php]: Check if $this->work_relation_ids is empty. Both for salary ids and employee ids.

File: modules/altinnsalary/view/salarylist.php [hunk 7]
```diff
@@ -162,13 +167,14 @@
             print $ready_for_altinn_status;
           } else {
         ?>
-          <a class='button' href="<? print $_lib['sess']->dispatch ?>t=altinnsalary.list&altinnReport1_periode=<? print $_periode; ?>&only_register_employee=1&use_employee[<? print $employee->AccountPlanID; ?>]=1&action_soap1=1">Register ansatte i Altinn</a>
+          <a class='button' href="<? print $_lib['sess']->dispatch ?>t=altinnsalary.list&altinnReport1_periode=<? print $_periode; ?>&only_register_employee=1&use_work_relation[<? print $work_relation->WorkRelationID; ?>]=1&action_soap1=1">Register ansatte i Altinn</a>
```
Comment (stoivo) [inline modules/altinnsalary/view/salarylist.php]: Register arbeidsforhold i Altinn

File: modules/salary/view/edit.php [hunk 8]
```diff
@@ -176,6 +176,39 @@
 ?>
 
 <? print $_lib['sess']->doctype ?>
+
+<? 
+  $wr_query = sprintf("SELECT
```
Comment (stoivo) [inline modules/salary/view/edit.php]: wr_query => work_relation_query

File: modules/salary/view/edit.php [hunk 9]
```diff
@@ -176,6 +176,39 @@
...
+                        w.OccupationID as OccupationID,
+                        w.WorkStart as WorkStart,
+                        w.WorkStop as WorkStop
+                      FROM 
+                        workrelation w
+                        LEFT JOIN subcompany sc ON w.SubcompanyID = sc.SubcompanyID
+                      GROUP BY 
+                        WorkRelationID
+                      ");
+  $result_set = $_lib['db']->db_query($wr_query);
```
Comment (stoivo) [inline modules/salary/view/edit.php]: result_set => work_relation_result

File: modules/salary/view/edit.php [hunk 10]
```diff
@@ -663,5 +718,58 @@
...
+    var selected_wr_id = <? print $head->WorkRelationID; ?>;
+    var out = "";
+    out += "<select name='salary_WorkRelationID_<? print $SalaryID; ?>' onchange='selected_work_relation = all_work_relations[selected_employee][this.value];'>"
+    out += "  <option value=''>Ikke valgt</option>";
+    for(var i in work_relations) {
+      is_selected = (work_relations[i]["WorkRelationID"] == selected_wr_id) ? ' selected ' : '';
+      out += "  <option value='" + work_relations[i]["WorkRelationID"] + "'" + is_selected + ">" + work_relations[i]["WorkRelationName"] + "</option>";
+    }
+    out += "</select>";
+    out += "<button onclick='updateInputValuesWithWorkRelationInfo(); return false;'>Oppdater arbeidsforhold</button>";
```
Comment (stoivo) [inline modules/salary/view/edit.php]: Oppdater => Endre

File: modules/salary/view/edit.php [hunk 11]
```diff
@@ -260,21 +300,36 @@
...
+      <div id="WorkTimeSchemeShowingValue"></div>
     </th>
     <th class="sub" colspan="3">
+      <div style="display: none;">
       <? print $_lib['form3']->Generic_menu3(array('data' => $_lib['form3']->_ALTINN['TypeOfEmploymentTypes'], 'table'=> 'salary', 'field'=>'TypeOfEmployment', 'pk'=>$head->SalaryID, 'value'=>$head->TypeOfEmployment, 'access' => $_lib['sess']->get_person('AccessLevel'), 'accesskey' => 'P', 'width' => 64)); ?>
+      </div>
+      <div id="TypeOfEmploymentShowingValue"></div>
     </th>
     <th class="sub" colspan="2">
+      <div style="display: none;">
```
Comment (stoivo) [inline modules/salary/view/edit.php]: Change this to dynamic loading into a hidden field.

File: modules/workrelation/model/workrelation.class.php [hunk 12]
```diff
@@ -0,0 +1,95 @@
+<?
+
+function validate_work_relations($array) {
```
Comment (stoivo) [inline modules/workrelation/model/workrelation.class.php]: array => work_relations

File: modules/workrelation/model/workrelation.class.php [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+      if($work_relation->InCurrentPositionSince > $work_relation->WorkStop && !empty_date($work_relation->WorkStop)) {
+        $errors[] = $error_prefix ."Samme posisjon siden dato kan ikke v&aelig;re etter slutt dato.";
+      }
+    }
+
+    // check for overlaps with other work relations
+    foreach ($array as $other_work_relation) {
+      if($work_relation->WorkRelationID == $other_work_relation->WorkRelationID) continue; // do not check with itself
+
+      if($work_relation->SubcompanyID == $other_work_relation->SubcompanyID) {
```
Comment (stoivo) [inline modules/workrelation/model/workrelation.class.php]: Change to 

``` php
if($work_relation->SubcompanyID != $other_work_relation->SubcompanyID) continue;
```

File: modules/workrelation/model/workrelation.class.php [hunk 14]
```diff
@@ -0,0 +1,95 @@
+<?
+
```
Comment (stoivo) [inline modules/workrelation/model/workrelation.class.php:4]: Wrap this in a class and make methods static.

**Comment by stoivo** [comment]
TODO:
- [x] Kommune should not be saved on WR
- [x] Kredittid oppdatert => Siste lønnsendrings dato
- [x] Kredittid oppdatert should be saved in WR
- [x] Save all fields for WR and kommune from employee in salary, when we generate altinn XML we should take it from salary if there is a salary.

File: modules/altinnsalary/model/altinnreport.class.php [hunk 15]
```diff
@@ -588,8 +588,10 @@ function queryStringForCurrentlyEmployedEmployees() {
  */
   function queryStringForIncludedWorkRelations() {
     $query_work_relations = "SELECT wr.*
-                             FROM workrelation wr
-                             WHERE wr.WorkRelationID IN (" . implode($this->work_relation_ids, ', ') . ")";
+                             FROM workrelation wr";
+    if (!empty($this->work_relation_ids)) {
+      $query_work_relations .= " WHERE wr.WorkRelationID IN (" . implode($this->work_relation_ids, ', ') . ")";
```
Comment (stoivo) [inline modules/altinnsalary/model/altinnreport.class.php:632]: Should we return a query that will select all WR if  `$this->work_relation_ids` is not set?

Comment (mladjo2505) [inline modules/altinnsalary/model/altinnreport.class.php:632]: Fixed

File: modules/accountplan/view/employee.php [hunk 16]
```diff
@@ -442,7 +442,7 @@
     <td><? print $_lib['form3']->text(array('table'=>$db_table2, 'field'=>'WorkPercent', 'value'=>$work_relation->WorkPercent, 'pk'=> $WorkRelationID, 'class'=>'lodoreqfelt')) ?></td>
     <td><? print $_lib['form3']->date(array('table'=>$db_table2, 'field'=>'WorkPercentUpdatedAt', 'value'=>$work_relation->WorkPercentUpdatedAt, 'pk'=> $WorkRelationID, 'class'=>'lodoreqfelt')) ?></td>
     <td><? print $_lib['form3']->text(array('table'=>$db_table2, 'field'=>'WorkMeasurement', 'value'=>$work_relation->WorkMeasurement, 'pk'=> $WorkRelationID, 'class'=>'lodoreqfelt')) ?></td>
-    <td><? print $_lib['form3']->kommune_menu(array('table'=>$db_table2, 'field'=>'KommuneID', 'value'=>$work_relation->KommuneID, 'pk'=> $WorkRelationID, 'class'=>'lodoreqfelt')) ?></td>
+    <td><? print $_lib['form3']->date(array('table'=>$db_table2, 'field'=>'SalaryDateChangedAt', 'value'=>$work_relation->SalaryDateChangedAt, 'pk'=> $WorkRelationID, 'class'=> 'lodoreqfelt')) ?></td>
```
Comment (mladjo2505) [inline modules/accountplan/view/employee.php]: @MrdjanStojimirovic : remove extra space

File: modules/salary/view/edit.php [hunk 17]
```diff
@@ -720,28 +708,36 @@
...
-    var options = document.getElementById(id).options;
-    for (var i = 0; i < options.length; i++) {
-      if (options[i].value == value) {              
-        options[i].selected = true;
-        break;
-      }
-    }
+  var all_work_relations = <? print json_encode($work_relations); ?>;
+  var selected_employee = <? print $head->AccountPlanID; ?>;
+  var selected_wr_id = <? print $head->WorkRelationID; ?>;
```
Comment (mladjo2505) [inline modules/salary/view/edit.php]: @MrdjanStojimirovic : refactor to `selected_work_relation_id`

**Comment by MrdjanStojimirovic** [comment]
First round review: @MrdjanStojimirovic
Final review with: @stoivo

File: code/lib/form/form_elements3.class.php [hunk 18]
```diff
@@ -1133,11 +1133,12 @@ function accountplan_number_menu($args) {
 
         if(isset($args['onchange'])) {
             $onchange = $args['onchange'];
```
Comment (stoivo) [inline code/lib/form/form_elements3.class.php]: Remove.

File: modules/workrelation/model/workrelation.class.php [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+          }
+        }
+        if($overlaps) $errors[] = $error_prefix ."Dato overlaper i arbeidsforhold ". $other_work_relation->WorkRelationID .".";
+      }
+
+      if(self::empty_field($work_relation->OccupationID)) {
+        $errors[] = $error_prefix ."Yrke kan ikke v&aelig;re blank.";
+      }
+
+      if(self::empty_field($work_relation->KommuneID)) {
```
Comment (mladjo2505) [inline modules/workrelation/model/workrelation.class.php]: Remove

File: db/changes/128_add_work_relation.sql [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+WorkRelationID int(11) NOT NULL,
+PRIMARY KEY (AltinnReport1ID, WorkRelationID)
+);
+
+-- Migrate existing information from altinnReport1employee to
+-- altinnReport1WorkRelation table linking the newly created working relation to
+-- the employee it belongs to
+INSERT INTO altinnReport1WorkRelation(AltinnReport1ID, WorkRelationID)
+SELECT ar1e.AltinnReport1ID, wr.WorkRelationID
+FROM altinnReport1employee ar1e JOIN accountplan a ON ar1e.AccountPlanID = a.AccountPlanID JOIN workrelation wr ON wr.AccountPlanID = a.AccountPlanID ORDER BY wr.WorkRelationID LIMIT 1;
```
Comment (mladjo2505) [inline db/changes/128_add_work_relation.sql]: Remove `LIMIT 1`

**Pull request merged**